### PR TITLE
Finish transport adapters

### PR DESCRIPTION
**Pull request by jwhonce**
* Add support for ssh:// and tcp: transports
* Add minimal tests for each transport
* Update tests to reflect transport changes
* Cleaned up timeout handling, lower APIs do not test for None vs.
  non-existing keyword argument
* Introduced cached_property decorator, mostly to cache creating
  managers.
* Improved docstrings
* Improved formatting URLs for future SSL verification and
  authenication.
* Bump to __version__ 3.1.2.1 tracking to Podman
* Added pyxdg dependency

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

**Comment by jwhonce** [comment]
@mwhahaha PTAL

**Comment by jwhonce** [comment]
@cevich I assume, the ssh tests are going to fail...

File: podman/api/adapter_utils.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+def _key_normalizer(key_class: NamedTuple, request_context: Mapping) -> Mapping:
+    """Create a pool key out of a request context dictionary.
+
+    According to RFC 3986, both the scheme and host are case-insensitive.
+    Therefore, this function normalizes both before constructing the pool
+    key for an HTTPS request. If you wish to change this behaviour, provide
+    alternate callables to ``key_fn_by_scheme``.
+
+    Args:
+        key_class: The class to use when constructing the key. This should be a namedtuple
```
Comment (TomSweeneyRedHat) [inline podman/api/adapter_utils.py:14]:
```suggestion
        key_class: The class to use when constructing the key. This should be a named tuple
```

File: podman/api/adapter_utils.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            with the scheme and host keys at a minimum.
+        request_context: An object that contain the context for a request.
+
+    Returns:
+        A namedtuple that can be used as a connection pool key.
```
Comment (TomSweeneyRedHat) [inline podman/api/adapter_utils.py:19]:
```suggestion
        A named tuple that can be used as a connection pool key.
```

File: podman/api/adapter_utils.py (hunk 1 above)
Comment (TomSweeneyRedHat) [reply podman/api/adapter_utils.py:14]: Ignore, just realized it's a Python data type

**Comment by TomSweeneyRedHat** [comment]
@jwhonce unfortunately the right call on the tests.

**Comment by cevich** [comment]
@jwhonce I don't think this is correct: `ssh ... ssh://root@localhost:22`.  I think what you intend is `ssh ... -p 22 root@localhost`

File: podman/api/ssh.py [hunk 3]
```diff
@@ -1,75 +1,300 @@
...
+            "-N",
+            "-L",
+            f"{self.local_sock}:{uri.path}",
+        ]
+
+        if self.identity is not None:
+            path = pathlib.Path(self.identity).expanduser()
+            command += ["-i", str(path)]
+
+        command += [f"ssh://{uri.netloc}"]
```
Comment (cevich) [inline podman/api/ssh.py:75]: here.  I think you need to break out the port number into a `-p` option.  Also I'm not sure about using the `ssh://` prefix...I've never used that before (maybe it works?).

Comment (cevich) [reply podman/api/ssh.py:75]: Oh okay, I see in the man page now:

> ssh connects and logs into the specified destination, which may be specified
> as either [user@]hostname or a URI of the form ssh://[user@]hostname[:port]

Hrmmmm.

**Pull request merged**